pkg/decoder/decoder.go

```diff
@@ -84,25 +136,105 @@ func (d *Decoder) sendBuffuredMessage(offset int64) {
 	d.msgBuffer.Reset()
 }
 
-// decodeIncomingData splits raw data based on `\n`, creates and sends messages to a channel
-func (d *Decoder) decodeIncomingData(inBuf []byte, offset int64) {
+// processBufferedLine checks the new line, appends the whole line or just a piece to the message
+// and sends the message if needed
+func (d *Decoder) processNewLine(offset int64) {
+	newLine := make([]byte, d.lineBuffer.Len())
+	copy(newLine, d.lineBuffer.Bytes())
+	defer d.lineBuffer.Reset()
+
+	if d.singleline {
+		d.msgBuffer.Write(newLine)
+		d.sendBufferedMessage(offset)
+		return
+	}
+
+	if d.msgBuffer.Len() == 0 {
+		d.msgBuffer.Write(newLine)
+		return
+	}
+
+	if d.re.Match(newLine) {
+		d.sendBufferedMessage(offset - int64(1+len(newLine))) // remove '\n' + new line length
+		d.msgBuffer.Write(newLine)
+		return
+	}
+
+	maxLen := maxMessageLen - d.msgBuffer.Len() - 2 
+	if len(newLine) >= maxLen {
+		d.msgBuffer.Write([]byte(`\n`))
+		d.msgBuffer.Write(newLine[:maxLen])
+		d.msgBuffer.Write(truncatedMsg)
+		d.sendBufferedMessage(offset - int64(1+len(newLine[maxLen:]))) // remove '\n' + the length of the piece of the line that can't fit
+		d.msgBuffer.Write(truncatedMsg)
+		d.msgBuffer.Write(newLine[maxLen:])
+		return
+	}
+
+	// append a new line to the message
+	d.msgBuffer.Write([]byte(`\n`))
+	d.msgBuffer.Write(newLine)
+}
+
+// recoverTooLongBufferedLine truncates the new line and sends its left part and the message if needed
+func (d *Decoder) recoverTooLongBufferedLine(offset int64) {
```

Comment:
1. This logic looks like code duplication, could it be refactored?
2. I think this is still too complex. What about the following behavior:
- `decodeIncomingData`: splits on `\n`, if line is too long (`j == maxj`), it calls `processNewLine` with an extra param (lineTooLong)
- `processNewLine`: when called with `lineTooLong=true`, adds TRUNCATED and flushes current message, as well as new message
- `processNewLine`: when called with `lineTooLong=false`, counts its buffer size, and adds TRUNCATED if needed. otherwise keeps same logic